### PR TITLE
User model modularize Location and InAppMessages

### PR DESCRIPTION
**Pull request by emawby**
# Description
## One Line Summary
Move location and InAppMessages to their own framework

## Details
- Create separate frameworks for location and InAppMessages
- OneSignalFramework will not directly depend on these relying on the App to include them.
- The public api protocols for these have been moved to OneSignalCore along with stub implementations that will be used if the frameworks are not included.
- This is currently using NSClassForString along with performSelector/NSInvocation to call into the libraries from the OneSignalFramework. This could be cleaned up in the future to use dependency injection instead.

### Distribution
- For Swift Package Manager we will create InAppMessaging and Location products
- For Cocoapods we will create InAppMessaging and Location subspecs and not have the OneSignal spec depend on those subspecs. Then we should update our cocoapods instructions to **NOT use Cocoapods for your NotificationServiceExtension target**. This is due to the [duplicate output issue](https://github.com/CocoaPods/CocoaPods/issues/11263) Cocoapods has. Instead customers can use OneSignalExtension by either adding the XCFramework directly in the libraries and frameworks project section, or have their runpaths look for the framework in the app target. I need to test this strategy fully with an app store upload, but it has been successful through the archive step.

### Motivation
Modularize IAMs and Location so that customers can exclude them from their apps. This will help with app store rejections due to these features.

### Scope
- InAppMessages: This code has moved to its own framework. It also has an indirect dependency on OneSignalLocation and uses NSInvocation to prompt for location.
- Location: This code has moved to its own framework.
- Migrations: In App Message migrations have been moved to the OneSignalInAppMessages framework.
- Lifecycle controller: This class calls location `onFocus` and InAppMessages `appDidBecomeActive`
- Sessions: Moved session time to `OSSessionManager` from `OneSignal`

### OPTIONAL - Other
**OPTIONAL -** Feel free to add any other sections or sub-sections that can explain your PR better.

# Testing
## Unit testing
**OPTIONAL -**  Explain unit tests added, if not clear in the code.

## Manual testing
**RECOMMEND - OPTIONAL -** Explain what scenarios were tested and the environment.
Example: Tested opening a notification while the app was foregrounded, app build with Android Studio 2020.3 with a fresh install of the OneSignal example app on a Pixel 6 with Android 12.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [ ] Public API changes

# Checklist
## Overview
   - [ ] I have filled out all **REQUIRED** sections above
   - [ ] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [ ] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [ ] I have included test coverage for these changes, or explained why they are not needed
   - [ ] All automated tests pass, or I explained why that is not possible
   - [ ] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [ ] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [ ] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-iOS-SDK/1273)

### CODE REVIEW

**Review by jkasten2 — CHANGES_REQUESTED** [review]
Code changes themselves look good, however there are a number of wrong copyrights with "Hiptic" on new files that need to be corrected.
Branch also contained rebuilt frameworks which made it harder to read what changed.

File: iOS_SDK/OneSignalSDK/OneSignalCore/Source/OSStubLocation.m [hunk 1]
```diff
@@ -0,0 +1,32 @@
+//
+//  OSStubLocation.m
+//  OneSignalCore
+//
+//  Created by Elliot Mawby on 6/21/23.
+//  Copyright © 2023 Hiptic. All rights reserved.
```
[the remark this answers is not in the document]
Comment (emawby) [reply iOS_SDK/OneSignalSDK/OneSignalCore/Source/OSStubLocation.m]: Done!

File: iOS_SDK/OneSignalSDK/OneSignal.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -4388,39 +4628,394 @@
...
+				DEFINES_MODULE = YES;
+				DEVELOPMENT_TEAM = 99SW8E36CT;
+				DYLIB_COMPATIBILITY_VERSION = 1;
+				DYLIB_CURRENT_VERSION = 1;
+				DYLIB_INSTALL_NAME_BASE = "@rpath";
+				ENABLE_MODULE_VERIFIER = YES;
+				GCC_C_LANGUAGE_STANDARD = gnu11;
+				GENERATE_INFOPLIST_FILE = YES;
+				INFOPLIST_FILE = OneSignalLocationFramework/Info.plist;
+				INFOPLIST_KEY_NSHumanReadableCopyright = "Copyright © 2023 Hiptic. All rights reserved.";
```
[the remark this answers is not in the document]
Comment (emawby) [reply iOS_SDK/OneSignalSDK/OneSignal.xcodeproj/project.pbxproj]: Done!

File: iOS_SDK/OneSignalSDK/OneSignal.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -4388,39 +4628,394 @@
...
+				GCC_C_LANGUAGE_STANDARD = gnu11;
+				GCC_DYNAMIC_NO_PIC = NO;
+				GCC_OPTIMIZATION_LEVEL = 0;
+				GCC_PREPROCESSOR_DEFINITIONS = (
+					"DEBUG=1",
+					"$(inherited)",
+				);
+				GENERATE_INFOPLIST_FILE = YES;
+				INFOPLIST_FILE = OneSignalLocationFramework/Info.plist;
+				INFOPLIST_KEY_NSHumanReadableCopyright = "Copyright © 2023 Hiptic. All rights reserved.";
```
[the remark this answers is not in the document]
Comment (emawby) [reply iOS_SDK/OneSignalSDK/OneSignal.xcodeproj/project.pbxproj]: Done!

**Review by jkasten2 — APPROVED** [review]

**Pull request merged**